### ISSUE TITLE
Banana long integer size limitation should be closer to the string length limitation

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4402]
|<img alt="Grashopa's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Grashopa reported|
|-|-|
|Trac ID|trac#4402|
|Type|defect|
|Created|2010-04-15 14:06:05Z|
|Branch|https://github.com/twisted/twisted/tree/spread-limits-4402|

Banana limits strings (and lists) to 640 * 1024 bytes (or elements) as protection against memory exhaustion issues.

Likewise, it limits long integers to around 2 ** 448.  It takes 65 bytes to encode the largest long integer.  It's not clear that there's any reason to limit the encoded size of an integer to something 4 orders of magnitude smaller than the limit on the encoded size of strings or lists.

<details><summary>Searchable metadata</summary>

```
trac-id__4402 4402
type__defect defect
reporter__Grashopa Grashopa
priority__normal normal
milestone__ 
branch__branches_spread_limits_4402 branches/spread-limits-4402
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1271340365000000 1271340365000000
changetime__1272465405000000 1272465405000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Grashopa's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|Grashopa commented|
|-|-|

Obviously we could also have a list of ints of any size as well.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set status to `closed`|
|-|-|

This is by design.  Accepting arbitrarily large _foo_ allows for memory exhaustion !DoS attacks.

**Comment by twisted-trac** [comment]
|<img alt="Grashopa's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|Grashopa set status to `reopened`|
|-|-|

Replying to [exarkun](#note_2):
> This is by design.  Accepting arbitrarily large _foo_ allows for memory exhaustion !DoS attacks.

This is no different than accepting a string size 640k. As I said simply duplicate the mechanism for passing a string when sending a long that surpasses whatever your current long size limit is.

**Comment by twisted-trac** [comment]
|<img alt="Grashopa's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|Grashopa commented|
|-|-|

Either way I'll be modifying banana.py for my own code when I get to it in a few months time. So if you'd like to consider the patch let me know.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Adjusting the ticket description to match.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Since there's `twisted.spread.banana.setPrefixLength`, I'm not sure this is really something that needs to be changed.  If you want to pass larger integers around, you should be able to just raise the prefix length with that function in your application code.

Perhaps the PB docs could do with a section on these limitations and how to overcome them, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [28874]) Branching to 'spread-limits-4402'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [28875]) words

refs [#4402](https://github.com/twisted/twisted/issues/4402)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst set owner to @exarkun|
|-|-|

1. The link to 'twisted.spread.banana.setPrefixLimit' should be an API link (I forget the magic syntax - class="API" or something). Maybe the constants too, though that matters less.

2. "Perspective Broker imposes some additional limits on top of these lower level limits." - really, only one limit sounds like, not multiple limits.

3. "PB Restrictions" may be better than limitations; this isn't abut missing functionality, but rather about things we chose to restrict.

Other than that, looks good.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [28876]) Replace "Limitations" with "Limits"; make setPrefixLimit an API link; change last limits section to use the singular, since it only talks about one limit

refs [#4402](https://github.com/twisted/twisted/issues/4402)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

Thanks!  All points addressed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst set owner to @exarkun|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [28877]) Merge spread-limits-4402

Author: exarkun
Reviewer: itamar
Fixes: [#4402](https://github.com/twisted/twisted/issues/4402)

Add some documentation describing certain arbitrary limits imposed by banana and
pb, and describing how they can be changed.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
